src/color_range.cpp

```diff
@@ -21,97 +21,101 @@
-	const uint16_t new_green = new_range.mid().g;
-	const uint16_t new_blue  = new_range.mid().b;
+#ifdef HAVE_CXX20
+#include <span>
+#endif
 
-	const uint16_t max_red   = new_range.max().r;
-	const uint16_t max_green = new_range.max().g;
-	const uint16_t max_blue  = new_range.max().b;
+namespace
+{
+#ifdef HAVE_CXX20
+std::vector<color_t> recolor_range_impl(const color_range& new_range, std::span<color_t> old_rgb)
+#else
+template<typename Container>
+std::vector<color_t> recolor_range_impl(const color_range& new_range, const Container& old_rgb)
+#endif
+{
+	std::vector<color_t> clist;
+	clist.reserve(old_rgb.size());
 
-	const uint16_t min_red   = new_range.min().r;
-	const uint16_t min_green = new_range.min().g;
-	const uint16_t min_blue  = new_range.min().b;
+	const color_t mid_c = new_range.mid();
+	const color_t max_c = new_range.max();
+	const color_t min_c = new_range.min();
 
 	// Map first color in vector to exact new color
-	const color_t temp_rgb = old_rgb.empty() ? color_t() : old_rgb[0];
+	const uint8_t reference_avg = old_rgb.empty()
+		? 255u
+		: (old_rgb[0].r + old_rgb[0].g + old_rgb[0].b) / 3;
 
-	const uint16_t reference_avg = (temp_rgb.r + temp_rgb.g + temp_rgb.b) / 3;
-
-	for(const auto& old_c : old_rgb) {
-		const uint16_t old_avg = (old_c.r + old_c.g + old_c.b) / 3;
-
-		// Calculate new color
-		uint32_t new_r = 0, new_g = 0, new_b = 0;
+	for(const color_t& old_c : old_rgb) {
+		const uint8_t old_avg = (old_c.r + old_c.g + old_c.b) / 3;
 
 		if(reference_avg && old_avg <= reference_avg) {
 			float old_ratio = static_cast<float>(old_avg) / reference_avg;
 
-			new_r = static_cast<uint32_t>(old_ratio * new_red   + (1 - old_ratio) * min_red);
-			new_g = static_cast<uint32_t>(old_ratio * new_green + (1 - old_ratio) * min_green);
-			new_b = static_cast<uint32_t>(old_ratio * new_blue  + (1 - old_ratio) * min_blue);
+			clist.emplace_back(
+				std::min<uint8_t>(255u, old_ratio * mid_c.r + (1 - old_ratio) * min_c.r),
+				std::min<uint8_t>(255u, old_ratio * mid_c.g + (1 - old_ratio) * min_c.g),
+				std::min<uint8_t>(255u, old_ratio * mid_c.b + (1 - old_ratio) * min_c.b)
+			);
 		} else if(reference_avg != 255) {
 			float old_ratio = (255.0f - static_cast<float>(old_avg)) / (255.0f - reference_avg);
 
-			new_r = static_cast<uint32_t>(old_ratio * new_red   + (1 - old_ratio) * max_red);
-			new_g = static_cast<uint32_t>(old_ratio * new_green + (1 - old_ratio) * max_green);
-			new_b = static_cast<uint32_t>(old_ratio * new_blue  + (1 - old_ratio) * max_blue);
+			clist.emplace_back(
+				std::min<uint8_t>(255u, old_ratio * mid_c.r + (1 - old_ratio) * max_c.r),
+				std::min<uint8_t>(255u, old_ratio * mid_c.g + (1 - old_ratio) * max_c.r),
+				std::min<uint8_t>(255u, old_ratio * mid_c.b + (1 - old_ratio) * max_c.r)
+			);
 		} else {
-			// Should never get here.
-			// Would imply old_avg > reference_avg = 255
+			// Should never get here. Would imply old_avg > reference_avg = 255
 			assert(false);
```

Comment:
While it's not something changing with this PR, is this something that really would need to make wesnoth exit?

Reply:
Forgot to address this. No, and it’s unreachable anyway, so I should just remove it.